### PR TITLE
DLNA PlayTo - implemented DLNA subscriptions

### PR DESCRIPTION
**Pull request by BaronGreenback**
Changes
-----
- DLNA events subscriptions for controls and playback used if the device supports them.
- Support for non-standard volume ranges on DLNA devices. (Not in the 0-100 range)
- PlayNext functionality added. : Inserts the track(s) next into the playlist. Starts playing if nothing in queue.
- PlayLast functionality added : Inserts the track(s) at the end of the playlist. Starts playing if nothing in queue.
- Shuffle functionality added : Shuffles the playlist and starts playing first track.
- Works with badly formed XML responses. (Sharp TV's)
- Progress bar adjusted for network lag between client and server.
- Client request queuing so the same command won't be sent multiple times to the client.
- Client device changed are reflected within the web environment. (eg. seeking whilst playing on a phone, updates the position in the web browser, etc.)
- Device Session initialisation occurs at first action, not at first detection of player.
- Centralised I/O functionality.
- Device communications rate-limited to stop devices being overloaded.
- Extensive error handling around all functionality as I was encountering multiple exceptions with the previous code. The system should now be able to recover from any error, including invalid xml responses from the clients.
- Extensive debug logging.
- Documented.

Tested on:
---
Hisense TV
Sharp TV
BubbleUPnP for DLNA

Depends on: PR #3542

Logging
----
To enable debug logging, add this the override section of your logging.default.json file.

	"Emby.Dlna": "Debug",
	"Emby.Dlna.Eventing": "Debug"

### CODE REVIEW

**Comment by BaronGreenback** [comment]
See https://github.com/jellyfin/jellyfin-web/issues/1540 for volume control issue.

File: Emby.Dlna/PlayTo/PlayToManager.cs [hunk 1]
```diff
@@ -64,11 +76,18 @@ public PlayToManager(ILogger logger, ISessionManager sessionManager, ILibraryMan
...
         }
 
+        public event EventHandler<DlnaEventArgs> DLNAEvents;
+
         public void Start()
         {
             _deviceDiscovery.DeviceDiscovered += OnDeviceDiscoveryDeviceDiscovered;
         }
 
+        public Task FireEvent(DlnaEventArgs args)
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/PlayToManager.cs:86]: Method to pass info posted to web service through to the device.

File: SsdpDeviceLocator.cs [hunk 2]
```diff
@@ -0,0 +1,636 @@
+using System;
```
Comment (BaronGreenback) [inline SsdpDeviceLocator.cs:1]: Nothing changed in this file - I removed a separate PR

File: Emby.Dlna/PlayTo/TransportCommands.cs [hunk 3]
```diff
@@ -2,6 +2,7 @@
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/TransportCommands.cs:2]: Refactoring and comments only.

File: Emby.Dlna/PlayTo/DeviceInfo.cs [hunk 4]
```diff
@@ -40,6 +40,7 @@ public string BaseUrl
             set => _baseUrl = value;
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/DeviceInfo.cs:40]: Comment only. No other change.

File: Emby.Server.Implementations/HttpClientManager/HttpClientManager.cs [hunk 5]
```diff
@@ -319,15 +319,17 @@ private async Task EnsureSuccessStatusCode(HttpResponseMessage response, HttpReq
             {
                 return;
             }
-
+            
+            string msg = string.Empty;
             if (options.LogErrorResponseBody)
             {
-                string msg = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                msg = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
```
Comment (BaronGreenback) [inline Emby.Server.Implementations/HttpClientManager/HttpClientManager.cs:326]: This is PR #3542

File: Emby.Dlna/PlayTo/XMLUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,259 @@
+#pragma warning disable IDE0059 // Unnecessary assignment of a value : Left in to aid in debugging.
+
+using System;
+using System.Text;
+using System.Web;
+using XMLProperties = System.Collections.Generic.Dictionary<string, string>;
+
+namespace Emby.Dlna.PlayTo
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/XMLUtilities.cs:8]: A XML parser that parses "invalid" XML code.

Whoever programmed Sharp TV's implementation of PlayTo couldn't be bothered to ensure the XML was valid, so HTML encoded it and wrapped it in a XML shell.

File: Emby.Dlna/PlayTo/PlayToController.cs [hunk 7]
```diff
@@ -345,65 +372,117 @@ public Task SendPlayCommand(PlayRequest command, CancellationToken cancellationT
...
+                    _session.RemoteEndPoint,
+                    user);
             }
 
-            return PlayItems(playlist, cancellationToken);
+            return PlayItems();
         }
 
-        private Task SendPlaystateCommand(PlaystateRequest command, CancellationToken cancellationToken)
+        private Task SendPlaystateCommand(PlaystateRequest command)
```
Comment (BaronGreenback) [inline Emby.Dlna/PlayTo/PlayToController.cs:461]: Actions are queued, so no cancellationToken is required.

**Comment by andekande** [comment]
As this PR does not seem to be merged, are the proposed changes included in some other PR ?

**Comment by BaronGreenback** [comment]
@andekande There is currently a staged roll out of PR's that will have this included.

**Pull request closed without merging**